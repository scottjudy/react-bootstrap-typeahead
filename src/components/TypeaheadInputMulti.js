// @flow

/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */

import cx from 'classnames';
import React, { type Node } from 'react';

import Hint, { type ShouldSelect } from './Hint';
import Input from './Input';

import { isSelectable } from '../utils';
import withClassNames from '../behaviors/classNames';

import { BACKSPACE } from '../constants';

import type { InputProps, Option, RefCallback, ReferenceElement } from '../types';

type Props = {
  ...InputProps,
  children: Node,
  referenceElementRef: RefCallback<ReferenceElement>,
  selected: Option[],
  shouldSelectHint?: ShouldSelect,
};

class TypeaheadInputMulti extends React.Component<Props> {
  wrapperRef = React.createRef<HTMLDivElement>();
  _input: ?HTMLInputElement;

  render() {
    const {
      children,
      className,
      inputClassName,
      inputRef,
      placeholder,
      referenceElementRef,
      selected,
      shouldSelectHint,
      ...props
    } = this.props;

    return (
      <div
        className={cx('rbt-input-multi', className)}
        disabled={props.disabled}
        onClick={this._handleContainerClickOrFocus}
        onFocus={this._handleContainerClickOrFocus}
        ref={referenceElementRef}
        tabIndex={-1}>
        <div className="rbt-input-wrapper" ref={this.wrapperRef}>
          {children}
          <Hint shouldSelect={shouldSelectHint}>
            <Input
              {...props}
              className={inputClassName}
              onKeyDown={this._handleKeyDown}
              placeholder={selected.length ? '' : placeholder}
              ref={this.getInputRef}
              style={{
                backgroundColor: 'transparent',
                border: 0,
                boxShadow: 'none',
                cursor: 'inherit',
                outline: 'none',
                padding: 0,
                width: '100%',
                zIndex: 1,
              }}
            />
          </Hint>
        </div>
      </div>
    );
  }

  getInputRef = (input: ?HTMLInputElement) => {
    this._input = input;
    this.props.inputRef(input);
  }

  /**
   * Forward click or focus events on the container element to the input.
   */
  _handleContainerClickOrFocus = (e: SyntheticEvent<HTMLElement>) => {
    // Don't focus the input if it's disabled.
    if (this.props.disabled) {
      e.currentTarget.blur();
      return;
    }

    // Move cursor to the end if the user clicks outside the actual input.
    const inputNode = this._input;
    if (!inputNode) {
      return;
    }

    if (e.currentTarget !== inputNode && isSelectable(inputNode)) {
      inputNode.selectionStart = inputNode.value.length;
    }

    inputNode.focus();
  }

  _handleKeyDown = (e: SyntheticKeyboardEvent<HTMLInputElement>) => {
    const { onKeyDown, selected, value } = this.props;

    switch (e.keyCode) {
      case BACKSPACE:
        if (e.currentTarget === this._input && selected.length && !value) {
          // Prevent browser from going back.
          e.preventDefault();

          // If the input is selected and there is no text, focus the last
          // token when the user hits backspace.
          if (this.wrapperRef.current) {
            const { children } = this.wrapperRef.current;
            const lastToken = children[children.length - 2];
            lastToken && lastToken.focus();
          }
        }
        break;
      default:
        break;
    }

    onKeyDown(e);
  }
}

export default withClassNames(TypeaheadInputMulti);
