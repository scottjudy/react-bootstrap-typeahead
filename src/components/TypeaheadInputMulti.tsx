/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */

import cx from 'classnames';
import React, { FocusEvent, KeyboardEvent, MouseEvent, ReactNode } from 'react';

import Hint from './Hint';
import Input from './Input';

import { isSelectable } from '../utils';
import withClassNames from '../behaviors/classNames';

import { Option, RefElement, TypeaheadInputProps } from '../types';

interface Props extends TypeaheadInputProps {
  children: ReactNode;
  selected: Option[];
}

class TypeaheadInputMulti extends React.Component<Props> {
  wrapperRef = React.createRef<HTMLDivElement>();
  _input: RefElement<HTMLInputElement> = null;

  render() {
    const {
      children,
      className,
      inputClassName,
      inputRef,
      placeholder,
      referenceElementRef,
      selected,
      shouldSelectHint,
      ...props
    } = this.props;

    return (
      <div
        className={cx('rbt-input-multi', className)}
        onClick={this._handleContainerClickOrFocus}
        onFocus={this._handleContainerClickOrFocus}
        ref={referenceElementRef}
        tabIndex={-1}>
        <div className="rbt-input-wrapper" ref={this.wrapperRef}>
          {children}
          <Hint shouldSelect={shouldSelectHint}>
            <Input
              {...props}
              className={inputClassName}
              onClick={this._handleClick}
              onKeyDown={this._handleKeyDown}
              placeholder={selected.length ? '' : placeholder}
              ref={this.getInputRef}
              style={{
                backgroundColor: 'transparent',
                border: 0,
                boxShadow: 'none',
                cursor: 'inherit',
                outline: 'none',
                padding: 0,
                width: '100%',
                zIndex: 1,
              }}
            />
          </Hint>
        </div>
      </div>
    );
  }

  getInputRef = (input: RefElement<HTMLInputElement>) => {
    this._input = input;
    this.props.inputRef(input);
  };

  _handleClick = (e: MouseEvent<HTMLInputElement>) => {
    // Prevent clicks on the input from bubbling up to the container,
    // which then re-focuses the input.
    e.stopPropagation();
    this.props.onClick && this.props.onClick(e);
  };

  /**
   * Forward click or focus events on the container element to the input.
   */
  _handleContainerClickOrFocus = (
    e: MouseEvent<HTMLElement> | FocusEvent<HTMLElement>
  ) => {
    // Don't focus the input if it's disabled.
    if (this.props.disabled) {
      e.currentTarget.blur();
      return;
    }

    // Move cursor to the end if the user clicks outside the actual input.
    const inputNode = this._input;

    if (!inputNode) {
      return;
    }

    if (isSelectable(inputNode)) {
      inputNode.selectionStart = inputNode.value.length;
    }

    inputNode.focus();
  };

  _handleKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    const { onKeyDown, selected, value } = this.props;

    switch (e.key) {
      case 'Backspace':
        if (e.currentTarget === this._input && selected.length && !value) {
          // Prevent browser from going back.
          e.preventDefault();

          // If the input is selected and there is no text, focus the last
          // token when the user hits backspace.
          if (this.wrapperRef.current) {
            const { children } = this.wrapperRef.current;
            const lastToken = children[children.length - 2];
            lastToken && (lastToken as HTMLElement).focus();
          }
        }
        break;
      default:
        break;
    }

    onKeyDown && onKeyDown(e);
  };
}

export default withClassNames(TypeaheadInputMulti);
