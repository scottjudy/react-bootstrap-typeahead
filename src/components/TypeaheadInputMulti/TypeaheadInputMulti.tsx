/* eslint-disable jsx-a11y/no-static-element-interactions */
/* eslint-disable jsx-a11y/click-events-have-key-events */

import cx from 'classnames';
import React, { FocusEvent, KeyboardEvent, MouseEvent, ReactNode } from 'react';

import Hint from '../Hint';
import Input from '../Input';

import { isSelectable } from '../../utils';
import withClassNames from '../../behaviors/classNames';

import { Option, RefElement, TypeaheadInputProps } from '../../types';

export interface TypeaheadInputMultiProps extends TypeaheadInputProps {
  children: ReactNode;
  selected: Option[];
}

class TypeaheadInputMulti extends React.Component<TypeaheadInputMultiProps> {
  wrapperRef = React.createRef<HTMLDivElement>();
  _input: RefElement<HTMLInputElement> = null;

  render() {
    const {
      children,
      className,
      inputClassName,
      inputRef,
      placeholder,
      referenceElementRef,
      selected,
      ...props
    } = this.props;

    return (
      <div
        className={cx('rbt-input-multi', className)}
        onClick={this._handleContainerClickOrFocus}
        onFocus={this._handleContainerClickOrFocus}
        ref={referenceElementRef}
        tabIndex={-1}>
        <div className="rbt-input-wrapper" ref={this.wrapperRef}>
          {children}
          <Hint>
            <Input
              {...props}
              className={inputClassName}
              onKeyDown={this._handleKeyDown}
              placeholder={selected.length ? '' : placeholder}
              ref={this.getInputRef}
              style={{
                backgroundColor: 'transparent',
                border: 0,
                boxShadow: 'none',
                cursor: 'inherit',
                outline: 'none',
                padding: 0,
                width: '100%',
                zIndex: 1,
              }}
            />
          </Hint>
        </div>
      </div>
    );
  }

  getInputRef = (input: RefElement<HTMLInputElement>) => {
    this._input = input;
    this.props.inputRef(input);
  };

  /**
   * Forward click or focus events on the container element to the input.
   */
  _handleContainerClickOrFocus = (
    e: MouseEvent<HTMLElement> | FocusEvent<HTMLElement>
  ) => {
    // Don't focus the input if it's disabled.
    if (this.props.disabled) {
      e.currentTarget.blur();
      return;
    }

    const inputNode = this._input;
    if (
      !inputNode ||
      // Ignore if the clicked element is a child of the container, ie: a token
      // or the input itself.
      (e.currentTarget.contains(e.target as HTMLElement) &&
        e.currentTarget !== e.target)
    ) {
      return;
    }

    if (isSelectable(inputNode)) {
      // Move cursor to the end if the user clicks outside the actual input.
      inputNode.selectionStart = inputNode.value.length;
    }

    inputNode.focus();
  };

  _handleKeyDown = (e: KeyboardEvent<HTMLInputElement>) => {
    const { onKeyDown, selected, value } = this.props;

    if (e.key === 'Backspace' && selected.length && !value) {
      // Prevent browser from going back.
      e.preventDefault();

      // If the input is selected and there is no text, focus the last
      // token when the user hits backspace.
      if (this.wrapperRef.current) {
        const { children } = this.wrapperRef.current;
        const lastToken = children[children.length - 2];
        lastToken && (lastToken as HTMLElement).focus();
      }
    }

    onKeyDown && onKeyDown(e);
  };
}

export default withClassNames(TypeaheadInputMulti);
