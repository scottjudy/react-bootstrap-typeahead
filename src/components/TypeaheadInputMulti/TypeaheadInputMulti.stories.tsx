/* eslint-disable sort-keys,import/no-extraneous-dependencies */

import React, { ChangeEvent, useState } from 'react';
import { Story, Meta } from '@storybook/react';

import Token from '../Token';
import TypeaheadInputMulti, {
  TypeaheadInputMultiProps,
} from './TypeaheadInputMulti';

import options from '../../tests/data';
import { HintProvider, noop } from '../../tests/helpers';

export default {
  title: 'Components/TypeaheadInputMulti',
  component: TypeaheadInputMulti,
} as Meta;

const selected = options.slice(1, 4);

const defaultProps = {
  children: selected.map((option) => (
    <Token key={option.name} option={option} onRemove={noop}>
      {option.name}
    </Token>
  )),
  selected,
};

interface Args extends TypeaheadInputMultiProps {
  hintText?: string;
}

const Template: Story<Args> = ({ hintText = '', ...args }) => {
  const [value, setValue] = useState(args.value);
  const [inputNode, setInputNode] = useState<HTMLInputElement | null>(null);

  return (
    <HintProvider hintText={hintText} inputNode={inputNode}>
      <TypeaheadInputMulti
        {...args}
        inputRef={setInputNode}
        onChange={(e: ChangeEvent<HTMLInputElement>) =>
          setValue(e.target.value)
        }
        referenceElementRef={noop}
        value={value}
      />
    </HintProvider>
  );
};

export const Default = Template.bind({});
Default.args = {
  ...defaultProps,
};

export const WithHint = Template.bind({});
WithHint.args = {
  ...defaultProps,
  hintText: 'california',
  value: 'cali',
};

export const Disabled = Template.bind({});
Disabled.args = {
  ...defaultProps,
  children: selected.map((option) => (
    <Token disabled key={option.name} option={option} onRemove={noop}>
      {option.name}
    </Token>
  )),
  disabled: true,
};
