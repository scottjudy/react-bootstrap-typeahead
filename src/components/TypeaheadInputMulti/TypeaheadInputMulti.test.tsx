import React from 'react';

import * as stories from './TypeaheadInputMulti.stories';

import {
  composeStories,
  fireEvent,
  generateSnapshots,
  getInput,
  getTokens,
  render,
  userEvent,
} from '../../tests/helpers';

const { Default, Disabled } = composeStories(stories);

describe('<TypeaheadInputMulti>', () => {
  generateSnapshots(stories);

  it('renders a multi-select input with tokens', () => {
    const { container } = render(<Default />);
    // Find the token close buttons as a proxy for the token itself.
    expect(getTokens(container)).toHaveLength(3);
  });

  it('focuses the input', () => {
    const { container } = render(<Default />);

    // Test clicking the container, which forwards the click to the input.
    userEvent.click(container.firstElementChild);
    expect(getInput()).toHaveFocus();
  });

  it('does not focus a disabled input', () => {
    const { container } = render(<Disabled />);

    const input = getInput();
    expect(input).toBeDisabled();

    // Test clicking the container, which forwards the click to the input.
    userEvent.click(container.firstElementChild);
    expect(input).not.toHaveFocus();

    userEvent.click(input);
    expect(input).not.toHaveFocus();
  });

  it('prevents clicks on the input from bubbling', () => {
    const onClick = jest.fn();
    render(<Default onClick={onClick} value="calif" />);

    const input = getInput();

    input.focus();
    input.selectionStart = 2;
    // userEvent.click triggers the wrong behavior for some reason.
    fireEvent.click(input);

    // Cursor shouldn't move when the input is clicked once.
    expect(onClick).toHaveBeenCalledTimes(1);
    expect(input.selectionStart).toBe(2);
  });

  it('calls the keydown handler', () => {
    const onKeyDown = jest.fn();
    render(<Default onKeyDown={onKeyDown} />);

    getInput().focus();
    userEvent.keyboard('{enter}');

    expect(onKeyDown).toHaveBeenCalledTimes(1);
  });

  it('focuses the last token', () => {
    const { container } = render(<Default />);

    getInput().focus();
    userEvent.keyboard('{backspace}');

    const tokens = getTokens(container);
    const lastToken = tokens[tokens.length - 1];

    expect(lastToken).toHaveFocus();
  });

  it('does not focus the last token when the input has a value', () => {
    render(<Default value="foo" />);

    const input = getInput();
    input.focus();
    userEvent.keyboard('{backspace}');

    expect(input).toHaveFocus();
  });
});
