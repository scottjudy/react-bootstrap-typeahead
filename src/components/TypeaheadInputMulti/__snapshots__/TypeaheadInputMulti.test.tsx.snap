// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TypeaheadInputMulti> Default story renders snapshot 1`] = `
<div
  className="rbt-input-multi form-control rbt-input"
  onClick={[Function]}
  onFocus={[Function]}
  tabIndex={-1}
>
  <div
    className="rbt-input-wrapper"
  >
    <div
      className="rbt-token rbt-token-removeable"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      option={
        Object {
          "capital": "Juneau",
          "name": "Alaska",
          "population": 710249,
          "region": "West",
        }
      }
      tabIndex={0}
    >
      Alaska
      <button
        aria-label="Remove"
        className="close rbt-close rbt-token-remove-button"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
        <span
          className="sr-only visually-hidden"
        >
          Remove
        </span>
      </button>
    </div>
    <div
      className="rbt-token rbt-token-removeable"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      option={
        Object {
          "capital": "Phoenix",
          "name": "Arizona",
          "population": 6392307,
          "region": "West",
        }
      }
      tabIndex={0}
    >
      Arizona
      <button
        aria-label="Remove"
        className="close rbt-close rbt-token-remove-button"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
        <span
          className="sr-only visually-hidden"
        >
          Remove
        </span>
      </button>
    </div>
    <div
      className="rbt-token rbt-token-removeable"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      option={
        Object {
          "capital": "Little Rock",
          "name": "Arkansas",
          "population": 2915958,
          "region": "South",
        }
      }
      tabIndex={0}
    >
      Arkansas
      <button
        aria-label="Remove"
        className="close rbt-close rbt-token-remove-button"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
        <span
          className="sr-only visually-hidden"
        >
          Remove
        </span>
      </button>
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "flex": 1,
          "height": "100%",
          "position": "relative",
        }
      }
    >
      <input
        className="rbt-input-main"
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        placeholder=""
        style={
          Object {
            "backgroundColor": "transparent",
            "border": 0,
            "boxShadow": "none",
            "cursor": "inherit",
            "outline": "none",
            "padding": 0,
            "width": "100%",
            "zIndex": 1,
          }
        }
      />
      <input
        aria-hidden={true}
        className="rbt-input-hint"
        readOnly={true}
        style={
          Object {
            "backgroundColor": "transparent",
            "borderColor": "transparent",
            "boxShadow": "none",
            "color": "rgba(0, 0, 0, 0.54)",
            "left": 0,
            "pointerEvents": "none",
            "position": "absolute",
            "top": 0,
            "width": "100%",
          }
        }
        tabIndex={-1}
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`<TypeaheadInputMulti> Disabled story renders snapshot 1`] = `
<div
  className="rbt-input-multi form-control rbt-input"
  onClick={[Function]}
  onFocus={[Function]}
  tabIndex={-1}
>
  <div
    className="rbt-input-wrapper"
  >
    <div
      className="rbt-token rbt-token-disabled"
    >
      Alaska
    </div>
    <div
      className="rbt-token rbt-token-disabled"
    >
      Arizona
    </div>
    <div
      className="rbt-token rbt-token-disabled"
    >
      Arkansas
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "flex": 1,
          "height": "100%",
          "position": "relative",
        }
      }
    >
      <input
        className="rbt-input-main"
        disabled={true}
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        placeholder=""
        style={
          Object {
            "backgroundColor": "transparent",
            "border": 0,
            "boxShadow": "none",
            "cursor": "inherit",
            "outline": "none",
            "padding": 0,
            "width": "100%",
            "zIndex": 1,
          }
        }
      />
      <input
        aria-hidden={true}
        className="rbt-input-hint"
        readOnly={true}
        style={
          Object {
            "backgroundColor": "transparent",
            "borderColor": "transparent",
            "boxShadow": "none",
            "color": "rgba(0, 0, 0, 0.54)",
            "left": 0,
            "pointerEvents": "none",
            "position": "absolute",
            "top": 0,
            "width": "100%",
          }
        }
        tabIndex={-1}
        value=""
      />
    </div>
  </div>
</div>
`;

exports[`<TypeaheadInputMulti> WithHint story renders snapshot 1`] = `
<div
  className="rbt-input-multi form-control rbt-input"
  onClick={[Function]}
  onFocus={[Function]}
  tabIndex={-1}
>
  <div
    className="rbt-input-wrapper"
  >
    <div
      className="rbt-token rbt-token-removeable"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      option={
        Object {
          "capital": "Juneau",
          "name": "Alaska",
          "population": 710249,
          "region": "West",
        }
      }
      tabIndex={0}
    >
      Alaska
      <button
        aria-label="Remove"
        className="close rbt-close rbt-token-remove-button"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
        <span
          className="sr-only visually-hidden"
        >
          Remove
        </span>
      </button>
    </div>
    <div
      className="rbt-token rbt-token-removeable"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      option={
        Object {
          "capital": "Phoenix",
          "name": "Arizona",
          "population": 6392307,
          "region": "West",
        }
      }
      tabIndex={0}
    >
      Arizona
      <button
        aria-label="Remove"
        className="close rbt-close rbt-token-remove-button"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
        <span
          className="sr-only visually-hidden"
        >
          Remove
        </span>
      </button>
    </div>
    <div
      className="rbt-token rbt-token-removeable"
      onBlur={[Function]}
      onClick={[Function]}
      onFocus={[Function]}
      onKeyDown={[Function]}
      option={
        Object {
          "capital": "Little Rock",
          "name": "Arkansas",
          "population": 2915958,
          "region": "South",
        }
      }
      tabIndex={0}
    >
      Arkansas
      <button
        aria-label="Remove"
        className="close rbt-close rbt-token-remove-button"
        onClick={[Function]}
        onKeyDown={[Function]}
        tabIndex={-1}
        type="button"
      >
        <span
          aria-hidden="true"
        >
          ×
        </span>
        <span
          className="sr-only visually-hidden"
        >
          Remove
        </span>
      </button>
    </div>
    <div
      style={
        Object {
          "display": "flex",
          "flex": 1,
          "height": "100%",
          "position": "relative",
        }
      }
    >
      <input
        className="rbt-input-main"
        onChange={[Function]}
        onClick={[Function]}
        onKeyDown={[Function]}
        placeholder=""
        style={
          Object {
            "backgroundColor": "transparent",
            "border": 0,
            "boxShadow": "none",
            "cursor": "inherit",
            "outline": "none",
            "padding": 0,
            "width": "100%",
            "zIndex": 1,
          }
        }
        value="cali"
      />
      <input
        aria-hidden={true}
        className="rbt-input-hint"
        readOnly={true}
        style={
          Object {
            "backgroundColor": "transparent",
            "borderColor": "transparent",
            "boxShadow": "none",
            "color": "rgba(0, 0, 0, 0.54)",
            "left": 0,
            "pointerEvents": "none",
            "position": "absolute",
            "top": 0,
            "width": "100%",
          }
        }
        tabIndex={-1}
        value="california"
      />
    </div>
  </div>
</div>
`;
